### PR TITLE
WFCORE-950 - allow changing of --jboss-home in non-modular

### PR DESCRIPTION
**Pull request by luck3y**
The should help if a user enters an incorrect --jboss-home or restarts with a different --jboss-home.

### CODE REVIEW

**Comment by luck3y** [comment]
There appears to be an issue with running the full testsuite that I'm investigating as part of this. Running each suite separately is fine, but running everything together appears to result in failures in the domain suite after completion of manualmode.

**Comment by luck3y** [comment]
retest this please

File: embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java [hunk 1]
```diff
@@ -293,8 +296,14 @@ private static ModuleLoader setupModuleLoader(String modulePath, String... syste
             }
             WildFlySecurityManager.setPropertyPrivileged("jboss.modules.system.pkgs", packages.toString());
 
-            // Get the module loader
-            return Module.getBootModuleLoader();
+            // setup the module class so we can reload it and reset the static properties later if we need to
+            URL[] urls = new URL[] {new File(modulePath).toURI().toURL(), new File(modulePath + "/../jboss-modules.jar").toURI().toURL()};
```
Comment (bstansberry) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java]: The callers of this method should know where the root is. The EmbeddedServerFactory.create methods take separate params for jbossHomePath and modulePath, so the relationship between modulePath and jboss-modules.jar may not be fixed like this. But the relationship between jbossHomePath and jboss-modules.jar should be.

File: embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java [hunk 2]
```diff
@@ -293,8 +296,14 @@ private static ModuleLoader setupModuleLoader(String modulePath, String... syste
...
+            ClassLoader classloader = new URLClassLoader(urls, ClassLoader.getSystemClassLoader());
+
+            Class clazz = classloader.loadClass("org.jboss.modules.Module");
+            return (ModuleLoader)clazz.getMethod("getBootModuleLoader").invoke(clazz);
```
Comment (bstansberry) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java]: Nice.  You mentioned reflection and I envisioned something much more painful than this. :)

Comment (luck3y) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java]: Hackity hack ;)

Comment (bstansberry) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java]: I don't get how it works though. :) If the Module class is visible in the system classloader, URLClassLoader should be returning that as it's a parent first CL. And if its not visible and Module is loaded from urls[0], then how can the result of getBootModuleLoader be cast to the correct ModuleLoader?

**Comment by luck3y** [comment]
I need to do some final testing to verify, but I think the class loading resets the boot module loader. It behaves like that in the CLI, but I do need to do a final verification.

**Comment by luck3y** [comment]
@bstansberry So re-testing this I'm not seeing the same behavior as I had previously, which is odd. I'm taking another look at what's going on.

**Comment by bstansberry** [comment]
I put the fixme label on this so it doesn't accidentally get merged.

**Comment by n1hility** [comment]
retest this please (don't worry not trying to merge yet)

**Comment by n1hility** [comment]
retest this please

**Comment by luck3y** [comment]
This has been updated after additional debugging & testing to:

- verify that the modules directory exists before we try and start loading modues
- correct properties still pointing to the previous jboss-home when an embedded server is started in a different directory.

A couple of issues remain, I've mail wildfly-dev to ask about a couple of things:

- allowing modules to be re-initialized by some method, and whether this would be safe or not
- after an embedded server has been started, any subsequent servers continue to log to the same server.log.

If someone could review the properties reset, I'd appreciate it, as it wasn't completely clear to me that resetting them like this was entirely correct w.r.t. directory names that may have been configured differently elsewhere.

File: embedded/src/main/java/org/wildfly/core/embedded/EmbeddedHostControllerFactory.java [hunk 3]
```diff
@@ -444,8 +426,8 @@ private void exit() {
         private static HostControllerEnvironment createHostControllerEnvironment(File jbossHome, String[] cmdargs, String hostName, long startTime) {
             try {
                 // for SecurityActions.getSystemProperty("jboss.home.dir") in InstallationManagerService
-                System.setProperty("jboss.home.dir", jbossHome.getAbsolutePath());
-                WildFlySecurityManager.setPropertyPrivileged("jboss.home.dir", jbossHome.getAbsolutePath());
+                System.setProperty(HostControllerEnvironment.HOME_DIR, jbossHome.getAbsolutePath());
+                WildFlySecurityManager.setPropertyPrivileged(HostControllerEnvironment.HOME_DIR, jbossHome.getAbsolutePath());
```
Comment (jamezp) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedHostControllerFactory.java]: Not fully reviewing this, but just caught this in passing. We should likely remove the `System.setProperty()` version if we need to do it privileged.

Comment (luck3y) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedHostControllerFactory.java]: @jamezp Thanks, done.

File: embedded/src/main/java/org/wildfly/core/embedded/EmbeddedHostControllerFactory.java [hunk 4]
```diff
@@ -447,6 +446,8 @@ private static HostControllerEnvironment createHostControllerEnvironment(File jb
                 System.setProperty("jboss.home.dir", jbossHome.getAbsolutePath());
                 WildFlySecurityManager.setPropertyPrivileged("jboss.home.dir", jbossHome.getAbsolutePath());
 
+                WildFlySecurityManager.setPropertyPrivileged(HostControllerEnvironment.HOME_DIR, jbossHome.getAbsolutePath());
+
```
Comment (bstansberry) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedHostControllerFactory.java]: L446 and 447 should be removed as 449 does the same thing.

File: embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java [hunk 5]
```diff
@@ -338,4 +340,44 @@ private static void setupLoggingSystem(ModuleLoader moduleLoader) {
...
+                throw new RuntimeException("Unknown embedded server type: " + embeddedServerType);
+        }
+
+        WildFlySecurityManager.setPropertyPrivileged(SYSPROP_KEY_JBOSS_HOME_DIR, jbossHomeDir);
+        WildFlySecurityManager.setPropertyPrivileged(SYSPROP_KEY_JBOSS_BASE_DIR, jbossBaseDir);
+        WildFlySecurityManager.setPropertyPrivileged(SYSPROP_KEY_JBOSS_CONFIG_DIR, jbossBaseDir + File.separator + "configuration");
+        WildFlySecurityManager.setPropertyPrivileged(SYSPROP_KEY_JBOSS_DEPLOY_DIR, jbossBaseDir + File.separator + "data" + File.separator +"content");
+        WildFlySecurityManager.setPropertyPrivileged(SYSPROP_KEY_JBOSS_TEMP_DIR, jbossBaseDir + File.separator + "data" + File.separator + "tmp");
+        WildFlySecurityManager.setPropertyPrivileged(SYSPROP_KEY_JBOSS_LOG_DIR, jbossBaseDir + File.separator + "log");
+        WildFlySecurityManager.setPropertyPrivileged(SYSPROP_KEY_JBOSS_DATA_DIR, jbossBaseDir + File.separator + "data");
```
Comment (bstansberry) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java]: The jboss.server.xxx.dir properties should not be being set if the embedded process is a Host Controller. The HC equivalents are jboss.domain.xxx.dir.

File: embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java [hunk 6]
```diff
@@ -338,4 +340,44 @@ private static void setupLoggingSystem(ModuleLoader moduleLoader) {
...
+        Object serverImpl;
+        try {
+            Properties sysprops = WildFlySecurityManager.getSystemPropertiesPrivileged();
+            Map<String, String> sysenv = WildFlySecurityManager.getSystemEnvironmentPrivileged();
+            String[] args = cmdargs != null ? cmdargs : new String[0];
+            serverImpl = createServerMethod.invoke(null, jbossHomeDir, moduleLoader, sysprops, sysenv, args);
+        } catch (final InvocationTargetException ite) {
+            throw EmbeddedLogger.ROOT_LOGGER.cannotCreateStandaloneServer(ite.getCause(), createServerMethod);
+        } catch (final IllegalAccessException iae) {
+            throw EmbeddedLogger.ROOT_LOGGER.cannotCreateStandaloneServer(iae, createServerMethod);
```
Comment (bstansberry) [inline embedded/src/main/java/org/wildfly/core/embedded/EmbeddedServerFactory.java]: This is a pre-existing issue, but might as well correct it now. These error messages are not correct in the HC case.

**Comment by luck3y** [comment]
@bstansberry Thanks, I'll address these and resend right away.

**Pull request merged**